### PR TITLE
Updates for GeometryBasics refactor

### LINKED ISSUE #2181: Order of decomposes coordinates is not counter/clockwise
The `Point2f`s that `decompose(Point2f, Rect(0, 0, 1, 1))` returns are not sorted in a clockwise order (nor counter clockwise). This is not ideal because plotting a decomposed rectangle doesn't work:
```julia
poly(decompose(Point2f, Rect(0, 0, 1, 1)))
```
tmp
nor does plotting polygons with a hole:
```julia
using GeometryBasics
p = Polygon(decompose(Point2f, Rect(0, 0, 1, 1)), [decompose(Point2f, Circle(Point2f(0.5), 0.25))])
poly(p)
```
tmp

Only specifying the coordinates manually:
```julia
p = Polygon(Point2f[(0,0), (1,0), (1,1), (0,1)], [decompose(Point2f, Circle(Point2f(0.5), 0.25))])
```
worked:
tmp

### LINKED ISSUE #3561: Some triangles are not rendered
- [x] are you running newest version (version from docs) ?
- [x] can you reproduce the bug with a fresh environment ? (`]activate --temp; add Makie`)
- [x] What platform + GPU are you on?

Various platform info:
```
MacOS 14.2.1 (Apple M2)
Julia 1.10.0
GLMakie 0.9.5, GeometryBasics 0.4.9
```

Certain triangles - depending on the vertex ordering - are not visible when plotted with `GLMakie.mesh`.

For example, if the following is run interactively:
```julia
using GeometryBasics
using GLMakie
t_ok = Triangle(Point3(0, 0, 0), Point3(1, 0, 3), Point3(0, 3, 2))
t_bad = Triangle(Point3(1, 0, -1), Point3(-1, 0, 3), Point3(0, 3, 2))

mesh(t_ok)
```
Then a blue triangle is shown, as expected. 
image

But if instead `mesh(t_bad)` is called, then no triangle is visible.
image

I think it's something to do with vertex ordering, because:
```julia
t_permuted = Triangle(t_bad.points[2], t_bad.points[1], t_bad.points[3])
mesh(t_permuted)
```
then shows the expected triangle.
image

---

Tracing through what is happening inside the `mesh` call, the eventual problem seems to be that the following call to `decompose` in `GeometryBasics` return any faces (the first step at extracting the points works fine, though):
```julia
points = decompose(Point{3,Float32}, t_bad)  # This step is fine
decompose(GLTriangleFace, points)            # This returns an empty array
```

I'm not sure if the "bug" is in GLMakie for not re-ordering the vertices before calling the `decompose` function, hence I'm filing it here rather than in GeometryBasics -- apologies if this is incorrect.

### PR DESCRIPTION
**Pull request by ffreyer**
# Description

#### Requires:

- https://github.com/JuliaGeometry/GeometryBasics.jl/pull/219
- https://github.com/JuliaIO/MeshIO.jl/pull/97
- https://github.com/SimonDanisch/ShaderAbstractions.jl/pull/28
- https://github.com/JuliaGraphics/RadeonProRender.jl/pull/22

Version bumps:
- https://github.com/JuliaGraphics/FreeTypeAbstraction.jl/pull/90
- https://github.com/jkrumbiegel/GridLayoutBase.jl/pull/61
- https://github.com/Kolaru/MathTeXEngine.jl/pull/125
- https://github.com/JuliaGeometry/Packing.jl/pull/12
- https://github.com/JuliaGraphics/RadeonProRender.jl/pull/24

This is still in prototyping stage as I'm looking what things GeometryBasics is missing/should do.

#### TODO:

- [x] Makie
  - [x] compiles
  - [x] tests 
- [x] get GLMakie working 
  - [x] compile
  - [x] mesh(rect3d)
  - [x] meshscatter()
  - [x] poly(polygon)
  - [x] mesh(cat, color = texture) (with https://github.com/JuliaIO/MeshIO.jl/pull/96)
  - [x] tests pass
- [x] WGLMakie
  - [x] compiles
  - [x] tests pass
- [x] CairoMakie
  - [x] compiles
  - [x] tests pass
- [x] RPRMakie
  - [x] compiles
  - [x] tests pass 

All RPRMakie examples run with https://github.com/JuliaGraphics/RadeonProRender.jl/pull/22 except for:
- datashader-rpr (the code is broken regardless of this pr)
- material_x.jl (crashes julia)
- sea_cables.jl (fails to download resources)

I'm not familiar enough with RPRMakie to say they look correct with confidence, but most look good to me. I'm not sure what to make of `eat_topographie_sphere.jl` (that's not a sphere?) and "volume.jl", they might be wrong.

#### Test failures:

- FigureSpec step 5: This pr somehow generates a more transparent marker than the refimg, which is barely transparent. The code uses `(:red, 0.5)`, so this seems more correct to me here. I'm just confused why it's different here at all
- Scatter with FastPixel: red outline color looks more intense in 3D (maybe color change, maybe thickness change, maybe both
- Triplot of a constained...: a line shifted from one diagonal to the other
- ~~shading of catmesh changed, seams visible~~ fixed
- ~~CairoMakie Merged color Mesh change~~ but that's expected due to updates to vertex order
- ~~CairoMakie voxel clipping looks more odd because triangle faces have shifted orientation~~ fixed via #4397
- CairoMakie voxels have a small shift in face intersections
- GLMakie heatmap cells shifted a little
- "FEM mesh 3D" is expected to fail due to vertex index merging not running on load anymore (i.e. the example no longer duplicates vertices unnecessarily and thus the color mapping changed)

#### Issues:

- fixes #3561
- closes #2181 (2D primitives are counter clockwise now)
- #4267 moved down to mesh and is the same as #3705 now as far as I can tell

## Type of change

- [x] Breaking change

## Checklist

- [x] Added an entry in CHANGELOG.md (for new features and breaking changes)
- [x] Added or changed relevant sections in the documentation
- [ ] Added unit tests for new algorithms, conversion methods, etc.
- [ ] Added reference image tests for new plotting functions, recipes, visual options, etc.

### CODE REVIEW

**Comment by ffreyer** [comment]
The refactor now makes it easier to work with per-face attributes, so I added an example to the mesh docs:

```julia
using GeometryBasics

# Reduce quality of sphere
s = Tesselation(Sphere(Point3f(0), 1f0), 12)
ps = coordinates(s)
fs = faces(s)

# Use a FaceView to with a new set of faces which refer to one color per face.
# Each face must have the same length as the respective face in fs.
# (Using the same face type guarantees this)
FT = eltype(fs); N = length(fs)
cs = FaceView(rand(RGBf, N), [FT(i) for i in 1:N])

# generate normals per face (this creates a FaceView as well)
ns = face_normals(ps, fs)

# Create mesh
m = GeometryBasics.mesh(ps, fs, normal = ns, color = cs)

mesh(m)
```
image

(This still requires vertex duplication, it just happens automatically now.)

Related Issues: #1369, #3624, maybe #742

**Comment by ffreyer** [comment]
https://github.com/MakieOrg/Makie.jl/pull/4319/commits/a8a61208738a2496cd888f4e213ccde6722a798e:
- added support for setting `mipmap` via `ShaderAbstractions.Sampler` (together with a commit to ShaderAbstractions)
- fixed N0f8 color types in WGLMakie
- fixed `minfilter` getting overwritten when using a Sampler in GLMakie

Not really related to GeometryBasics but I didn't want to make another pr in ShaderAbstractions for a 2 line change.

File: CairoMakie/src/primitives.jl [hunk 1]
```diff
@@ -1253,10 +1237,18 @@ function draw_atomic(scene::Scene, screen::Screen, @nospecialize(primitive::Maki
...
+        valid = [is_visible(primitive.clip_planes[], p) for p in pos]
+        pos = pos[valid]
+        colors = colors[valid]
+    end
+
+    marker = GeometryBasics.clear_faceviews(normal_mesh(Rect3f(Point3f(-0.5), Vec3f(1))))
 
     # For correct z-ordering we need to be in view/camera or screen space
-    model = copy(primitive.model[])
+    model = copy(primitive.model[])::Mat4d
```
Comment (SimonDanisch) [inline CairoMakie/src/primitives.jl:1251]: It seems weird to copy an immutable? Not part of this PR, but still we may want to remove it
